### ISSUE TITLE
tranlog with async type has the risk of data loss when processor dead suddenly

### ISSUE THREAD

**Issue by weizijun** [elastic/elasticsearch#43915]
**Describe the feature**:

**Elasticsearch version** (`bin/elasticsearch --version`): 6.6.1

**Plugins installed**: [analysis-icu,analysis-ik]

**JVM version** (`java -version`): 11.0.2+9-LTS

**OS version** (`uname -a` if on a Unix-like system): 3.10.0-514.16.1.es01.x86_64 #1 SMP Wed Oct 17 23:14:35 CST 2018 x86_64 x86_64 x86_64 GNU/Linux

**Description of the problem including expected versus actual behavior**:

when es datanode process dead suddenly. It may cause data loss.
I found the loss reason.
indices's settings is  index.translog.durability=async.
when es datanode process dead suddenly. translog.ckp 's globalCheckpoint maybe behand translog tlog's real data.
when recovery, TranslogSnapshot.next() will return null when readOperations > globalCheckpoint.
This will result in loss of behind translog data even if the data is complete.
When Server down, file meybe not fsync or refresh to dsik, this maybe cause dataloss.
But Server is ok, process exception coredump, translog is complete. It can recovery full data.

Will repair translog's ckp file before loading globalCheckpoint? Or recovery the whole complete translog?

**Steps to reproduce**:

Please include a *minimal* but *complete* recreation of the problem, including
(e.g.) index creation, mappings, settings, query etc.  The easier you make for
us to reproduce it, the more likely that somebody will take the time to look at it.

 1. write data.
 2. kill -9 processor
 3. check translog.ckp and translog-x.tlog's last seqNo.

**Provide logs (if relevant)**:

```
translog.ckp
00000000: 3fd7 6c17 0363 6b70 0000 0003 0000 0000  ?.l..ckp........
00000010: 037d a3df 0000 60c8 0000 0000 0000 0003  .}....`.........
00000020: 0000 0000 0000 6aac 0000 0000 0000 cb73  ......j........s
00000030: 0000 0000 0000 cb73 0000 0000 0000 0001  .......s........
00000040: ffff ffff ffff fffe c028 93e8 0000 0000  .........(......
00000050: 0000 0000 e543 a1d4 0a                   .....C...

translog-3.tlog
......
038575d0: 3022 2c22 7072 654f 6666 7365 7422 3a22  0","preOffset":"
038575e0: 3330 3335 3030 3131 3332 3722 7d01 2063  30350011327"}. c
038575f0: 3061 3831 3031 3935 6431 3538 6362 3739  0a810195d158cb79
03857600: 6434 3934 3338 3631 3033 6330 3430 3300  d494386103c0403.
03857610: 0000 0000 0000 0001 00ff ffff ffff ffff  ................
03857620: ff00 0000 0000 00cb f000 0000 0000 0000  ................
03857630: 0186 069c 0c0a                           ......
```

**Comment by weizijun** [comment]
@ywelsch I see this PR https://github.com/elastic/elasticsearch/pull/43205. Can this PR help to solve this issue?

**Comment by ywelsch** [comment]
I'm not sure that I understand the question here. Can you explain what the unexpected behavior here is? When using `translog.durability` set to `async`, there is a risk of losing acknowledged writes ([as the documentation says](https://www.elastic.co/guide/en/elasticsearch/reference/current/index-modules-translog.html)). Is the confusing part perhaps that it only talks about hardware failure, whereas you also lose data in case of process crashes?

> when recovery, TranslogSnapshot.next() will return null when readOperations > globalCheckpoint.

not true, where do you see that? Do you mean this line? https://github.com/elastic/elasticsearch/blob/ca5771c1f4aeb37ff133540927a8c770c46693c3/server/src/main/java/org/elasticsearch/index/translog/TranslogSnapshot.java#L71

Which recovery do you mean? recovery from store (i.e. as a primary)?

**Comment by weizijun** [comment]
> Is the confusing part perhaps that it only talks about hardware failure, whereas you also lose data in case of process crashes?

It's not hardware failure. hardware is ok, but process is down.
I simulate this case by kill -9 es pid, in somecase maybe jvm core, or oom-killer will cause the same problem.


> Which recovery do you mean? recovery from store (i.e. as a primary)?

yeah, recovery from store. indices number_of_replicas=0, It's primary recovery.

**Comment by weizijun** [comment]
In Provide logs. translog.ckp's globalCheckpoint=0xcb73. But in translog-3.tlog,there are many log's seqNo > 0xcb73. in translog-3.tlog's last log, seqNo=0xcbf0. Recovery from translog lost the translog operation from seqNo 0xcb74 to 0xcbf0.

**Comment by dnhatn** [comment]
> In Provide logs. translog.ckp's globalCheckpoint=0xcb73. But in translog-3.tlog,there are many log's seqNo > 0xcb73. in translog-3.tlog's last log, seqNo=0xcbf0. Recovery from translog lost the translog operation from seqNo 0xcb74 to 0xcbf0.

We only recover operations that we have **[manually](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java#L377-L378)** fsynced and recorded in the checkpoint files. If some operations are flushed and fsynced by OS, we won't try to scan and recover those operations. We should never attempt to recover translog as it's the source of truth.

**Comment by weizijun** [comment]
> > In Provide logs. translog.ckp's globalCheckpoint=0xcb73. But in translog-3.tlog,there are many log's seqNo > 0xcb73. in translog-3.tlog's last log, seqNo=0xcbf0. Recovery from translog lost the translog operation from seqNo 0xcb74 to 0xcbf0.
> 
> We only recover operations that we have **[manually](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java#L377-L378)** fsynced and recorded in the checkpoint files. If some operations are flushed and fsynced by OS, we won't try to scan and recover those operations. We should never attempt to recover translog as it's the source of truth.

I know the mechanism. But If translog has more operations then checkpoint, these operations have the chance to recover rather than drop them. 
TranslogWriter.syncUpTo will save the checkpoint. but if settings index.translog.durability=async, default index.translog.sync_interval=5s, syncUpTo will call every 5 seconds. If before next call syncUpTo,  index some data, translog will add these operations. If hardware failure, these translog maybe lost. But if only process crash, like jvm crash https://github.com/elastic/elasticsearch/issues/41432,  these translog are fine, but the checkpoint is behind these translog.
I think before shard start, can check translog.ckp file?
if checkpoint is behind translog log file, reset the translog.ckp to the right checkpoint.

**Comment by weizijun** [comment]
We are set number_of_replicas=0  in the log use case. Because log data is to large as PB level. We consumer kafka, sink data to es. when es index is success, we submit the kafka offset. But when es process crash unexpected, it will cause data loss. even if es index is return success, and we submit the kafka offset.

**Comment by weizijun** [comment]
translog.log:
```
		     |<---lost data--->|
---------------------|-----------------|
                     ^                 ^
              globalCheckpoint        end
```

**Comment by ywelsch** [comment]
You're confusing checkpoint and globalcheckpoint. What you meant to draw was something like:

```
		     |<---lost data--->|
---------------------|-----------------|
                     ^                 ^
              checkpoint.offset        end
```

> when es index is success, we submit the kafka offset.

That's the real bug here. The API does not offer this guarantee with async translog syncing. Also, even if we were to change the system as you suggested (which has other risks), you would still experience data loss in case of a system crash.

@dnhatn I think we should update the docs, which make it currently look like you only lose data in case of hardware failure.

**Comment by weizijun** [comment]
> You're confusing checkpoint and globalcheckpoint. What you meant to draw was something like:
> 
> ```
> 		     |<---lost data--->|
> ---------------------|-----------------|
>                      ^                 ^
>               checkpoint.offset        end
> ```
> 
> > when es index is success, we submit the kafka offset.
> 
> That's the real bug here. The API does not offer this guarantee with async translog syncing. Also, even if we were to change the system as you suggested (which has other risks), you would still experience data loss in case of a system crash.
> 
> @dnhatn I think we should update the docs, which make it currently look like you only lose data in case of hardware failure.

yeah, If translog operation lost due to hardware failure or other unexpected exception, This the index.translog.durability=async mechanism. But if  translog file is ok, recovery can do best to recovery these  operation.

**Comment by dnhatn** [comment]
I merged #44020 to clarify the consequence of the translog async setting. I also discussed this with Yannick. We preferred not to proceed with this change as translog is crucial for the consistency and resiliency of the system.

We can continue this conversation in the [forums](https://discuss.elastic.co/). Thank you very much for your interest in Elasticsearch.
